router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
yeah, it looks like we want to pass the context all the way down to `transport.roundTrip`. `http.Transporter` doesn't scope dials to requests; if a cancel happens during a dial, `net.Conn` (wrapped in `dialRes`) is moved to `idleConn` (via `handlePendingDial`). So I think it's optimal to check for cancelations before dialing at the beginning of the for loop of `roundTrip`:
```go
for _, backend := range backends {
	select {
	case <-ctx.Done():
		return nil, errRequestCanceled
	default:
	}

	req.URL.Host = backend
	// ...
}
```